lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,260 @@
+    if (auto testEnable = clkGate.getTestEnable()) {
+      OpBuilder builder(clkGate.getContext());
+      builder.setInsertionPointAfter(clkGate);
+      enable =
+          builder.create<comb::OrOp>(clkGate->getLoc(), enable, testEnable);
+    }
+    return enable;
+  }
+
+  void runOnOperation() override;
+
+private:
+  InstanceGraph *graph;
+  DenseSet<Operation *> opsToErase;
+  mlir::MLIRContext *context;
+};
+
+void SinkClockGatesPass::runOnOperation() {
+  graph = &getAnalysis<circt::hw::InstanceGraph>();
+  DenseSet<HWInstanceLike> instancesWithGatedClk;
+  DenseSet<Operation *> opsWithGatedClock;
+  context = getOperation().getContext();
+
+  // Record all the instances that have clock_gate as an input. This will be
+  // used to push the clock_gate down through the hierarchy to its users.
+  // This also collapses consecutive clock gates.
+  auto findInstancesWithGatedClock = [&](HWModuleOp hwModule) {
+    hwModule->walk([&](ClockGateOp clkGate) {
+      bool canErase = true;
+      for (auto &use : clkGate->getUses()) {
+        auto *useOp = use.getOwner();
+        if (auto inst = dyn_cast<HWInstanceLike>(useOp))
+          instancesWithGatedClk.insert(inst);
+        else if (auto userGate = dyn_cast<ClockGateOp>(useOp)) {
+          collapseConsecutiveClockGates(clkGate, userGate);
+        } else {
+          canErase = false;
+          opsWithGatedClock.insert(useOp);
+        }
+      }
+      if (canErase)
+        opsToErase.insert(clkGate);
+    });
+  };
+  // Seed phase, find the instances to start the traversal from.
+  // These instances have a gated clock as input.
+  getOperation()->walk(
+      [&](hw::HWModuleOp hwModule) { findInstancesWithGatedClock(hwModule); });
+
+  // Transformation phase, push the clock_gate through the instances and find
+  // all the indirect instances that have a gated clock.
+
+  // Iterate until none of the instances have a gated clock as input.
+  while (!instancesWithGatedClk.empty()) {
+
+    // Remove an entry from the set.
+    auto entry = instancesWithGatedClk.begin();
+    HWInstanceLike inst = *entry;
+    instancesWithGatedClk.erase(entry);
+
+    // Now, find all the gated clocks that are input to the instance, and
+    // duplicate it at the referenced module.
+
+    SmallVector<HWModuleOp> refMods = getReferencedModules(inst);
+    // Index of the instance operands that are a gated clock.
+    SmallVector<unsigned> gatedClkPorts;
+    // The new enable ports that will be added to the instance.
+    SmallVector<std::pair<StringAttr, Value>> enablePorts;
+    // The instance may or maynot have argNames attribute.
+    SmallVector<StringAttr> argNames = getInstanceArgNames(inst);
+    // Iterate over all instance operands to find a gated clock.
+    for (auto [index, in] : llvm::enumerate(inst->getOperands())) {
+      if (!isa<seq::ClockType>(in.getType()))
+        continue;
+      auto clkGate = dyn_cast_or_null<ClockGateOp>(in.getDefiningOp());
+      if (!clkGate)
```

Comment:
This seems like a bit fragile since it could break analysis when there is wire between instance and clock.

Reply:
As you pointed out, this would require a fixed-point analysis to handle all the possible dataflow operations that can propagate clocks. I think such operations of passing a clock by a wire are very rare, and it might be fine to ignore them for this transformation. This transformation does not guarantee that all clock gates will be removed. I want to do some more experiments with this transformation enabled, and then based on the need, we can utilize a fixed-point analysis.